includes/class-salesforce.php

```diff
@@ -0,0 +1,365 @@
+	const SALESFORCE_CLIENT_SECRET = 'newspack_salesforce_client_secret';
+	const SALESFORCE_ACCESS_TOKEN  = 'newspack_salesforce_access_token';
+	const SALESFORCE_REFRESH_TOKEN = 'newspack_salesforce_refresh_token';
+	const SALESFORCE_INSTANCE_URL  = 'newspack_salesforce_instance_url';
+	const SALESFORCE_WEBHOOK_ID    = 'newspack_salesforce_webhook_id';
+	const DEFAULT_SETTINGS         = [
+		'client_id'     => '',
+		'client_secret' => '',
+		'access_token'  => '',
+		'refresh_token' => '',
+		'instance_url'  => '',
+		'webhook_id'    => '',
+	];
+
+	/**
+	 * Get the Salesforce settings.
+	 *
+	 * @return Array of Salesforce settings.
+	 */
+	public static function get_salesforce_settings() {
+		$settings = self::DEFAULT_SETTINGS;
+
+		$client_id = get_option( self::SALESFORCE_CLIENT_ID, 0 );
+		if ( ! empty( $client_id ) ) {
+			$settings['client_id'] = $client_id;
+		}
+		$client_secret = get_option( self::SALESFORCE_CLIENT_SECRET, 0 );
+		if ( ! empty( $client_secret ) ) {
+			$settings['client_secret'] = $client_secret;
+		}
+		$access_token = get_option( self::SALESFORCE_ACCESS_TOKEN, 0 );
+		if ( ! empty( $access_token ) ) {
+			$settings['access_token'] = $access_token;
+		}
+		$refresh_token = get_option( self::SALESFORCE_REFRESH_TOKEN, 0 );
+		if ( ! empty( $refresh_token ) ) {
+			$settings['refresh_token'] = $refresh_token;
+		}
+		$instance_url = get_option( self::SALESFORCE_INSTANCE_URL, 0 );
+		if ( ! empty( $instance_url ) ) {
+			$settings['instance_url'] = $instance_url;
+		}
+
+		return $settings;
+	}
+
+	/**
+	 * Set the Salesforce settings.
+	 *
+	 * @param array $args Array of settings info.
+	 * @return array Updated settings.
+	 */
+	public static function set_salesforce_settings( $args ) {
+		$defaults = self::DEFAULT_SETTINGS;
+		$update   = wp_parse_args( $args, $defaults );
+
+		if ( array_key_exists( 'client_id', $args ) ) {
+			update_option( self::SALESFORCE_CLIENT_ID, $update['client_id'] );
+		}
+
+		if ( array_key_exists( 'client_secret', $args ) ) {
+			update_option( self::SALESFORCE_CLIENT_SECRET, $update['client_secret'] );
+		}
+
+		if ( array_key_exists( 'access_token', $args ) ) {
+			update_option( self::SALESFORCE_ACCESS_TOKEN, $update['access_token'] );
+		}
+
+		if ( array_key_exists( 'refresh_token', $args ) ) {
+			update_option( self::SALESFORCE_REFRESH_TOKEN, $update['refresh_token'] );
+		}
+
+		if ( array_key_exists( 'instance_url', $args ) ) {
+			update_option( self::SALESFORCE_INSTANCE_URL, $update['instance_url'] );
+		}
+
+		// Get webhook id.
+		$webhook_id = get_option( self::SALESFORCE_WEBHOOK_ID, 0 );
+
+		// If we're establishing a new connection and don't have an existing webhook, create a new one.
+		if ( empty( $webhook_id ) && ! empty( $args['refresh_token'] ) ) {
+			self::create_webhook();
+		}
+
+		// If we're resetting, let's delete the existing webhook, if it exists.
+		if ( ! empty( $webhook_id ) && '' === $args['refresh_token'] ) {
```

Comment:
When the webhook is **triggered**, I'm getting `Undefined index: refresh_token` error on this line.

Reply:
Good catch—I now check `isSet( $args['refresh_token' ] )` first, so I think that should suppress the error.

Reply:
@adekbadek I don't see this error in my logs anymore—can you please confirm?